model/value.go

```diff
@@ -152,30 +95,56 @@ func (s Sample) MarshalJSON() ([]byte, error) {
 			Value:     s.Value,
 		},
 	}
-
 	return json.Marshal(&v)
 }
 
+type sampleHistogramPairPtr struct {
+	Timestamp Time
+	Histogram *SampleHistogram
+}
+
+func (s *sampleHistogramPairPtr) UnmarshalJSON(buf []byte) error {
+	tmp := []interface{}{&s.Timestamp, &s.Histogram}
+	wantLen := len(tmp)
+	if err := json.Unmarshal(buf, &tmp); err != nil {
+		return err
+	}
+	if gotLen := len(tmp); gotLen != wantLen {
+		return fmt.Errorf("wrong number of fields: %d != %d", gotLen, wantLen)
+	}
+	return nil
+}
+
 // UnmarshalJSON implements json.Unmarshaler.
 func (s *Sample) UnmarshalJSON(b []byte) error {
 	v := struct {
-		Metric Metric     `json:"metric"`
-		Value  SamplePair `json:"value"`
+		Metric    Metric                 `json:"metric"`
+		Value     SamplePair             `json:"value"`
+		Histogram sampleHistogramPairPtr `json:"histogram"`
```

Comment:
We could probably do this as well and check for `v.Histogram != nil`. But if this gives problems, the current implementation looks fine as well.
```suggestion
		Histogram *SampleHistogramPair `json:"histogram"`
```

Reply:
Yes, I think (and hope) that will work.

Reply:
Or maybe we can just call the current `sampleHistogramPairPtr` `SampleHistogramPair` and get rid of the current `SampleHistogramPair`? Does something not work just because `SampleHistogramPair` will now contain a pointer to a histogram rather than the histogram itself?

Reply:
I tried the first suggestion in a reverted commit but it results in a panic when reading JSONs of float samples, but not sure if I was doing it correctly. As for the second suggestion, I think that would work, but not sure if the change to a pointer everywhere (with added complexity) is worth it just for this use case, since besides unmarshalling Samples, we can assume that everytime we use SampleHistogramPair, it should contain an actual histogram and don't need to check for nil pointers. Perhaps there's a better way to write this unmarshalling method for Sample without this `sampleHistogramPairPtr`, just that my knowledge of unmarshalling is currently too basic to figure out how.

Reply:
If it is getting too complicated, we can keep it as it is

Reply:
OK, let's keep it as is, but add a TODO to find out if there is a way to get rid of `SampleHistogramPair` (and rename `sampleHistogramPairPtr` to `SampleHistogramPair`).